packages/vite/src/node/server/middlewares/indexHtml.ts

```diff
@@ -108,32 +114,53 @@ const devHtmlHook: IndexHtmlTransformHook = async (
   const { config, moduleGraph, watcher } = server!
   const base = config.base || '/'
 
+  let proxyModulePath: string
+  let proxyModuleUrl: string
+
+  const trailingSlash = htmlPath.endsWith('/')
+  if (!trailingSlash && fs.existsSync(filename)) {
+    proxyModulePath = htmlPath
+    proxyModuleUrl = base + htmlPath.slice(1)
+  } else {
+    // There are users of vite.transformIndexHtml calling it with url '/'
+    // for SSR integrations #7993, filename is root for this case
+    // A user may also use a valid name for a virtual html file
+    // Mark the path as virtual in both cases so sourcemaps aren't processed
+    // and ids are properly handled
+    const validPath = `${htmlPath}${trailingSlash ? 'index.html' : ''}`
+    proxyModulePath = `\0${validPath}`
+    proxyModuleUrl = `${VALID_ID_PREFIX}${NULL_BYTE_PLACEHOLDER}${validPath}`
+  }
+
   const s = new MagicString(html)
   let inlineModuleIndex = -1
-  const filePath = cleanUrl(htmlPath)
+  const proxyCacheUrl = cleanUrl(proxyModulePath).replace(
+    normalizePath(config.root),
+    ''
+  )
   const styleUrl: AssetNode[] = []
 
   const addInlineModule = (node: ElementNode, ext: 'js') => {
     inlineModuleIndex++
 
-    const url = filePath.replace(normalizePath(config.root), '')
-
     const contentNode = node.children[0] as TextNode
 
     const code = contentNode.content
-    const map = new MagicString(html)
-      .snip(contentNode.loc.start.offset, contentNode.loc.end.offset)
-      .generateMap({ hires: true })
-    map.sources = [filename]
-    map.file = filename
+
+    let map: SourceMapInput | undefined
+    if (!proxyModulePath.startsWith('\0')) {
+      map = new MagicString(html)
+        .snip(contentNode.loc.start.offset, contentNode.loc.end.offset)
+        .generateMap({ hires: true })
+      map.sources = [filename]
+      map.file = filename
```

Comment:
IIUC, if `/` filename is `/path-to/folder`, sourcemap file need to point to `base + htmlPath.slice(1)`(exist file)?